bigtable/google/cloud/bigtable/retry.py

```diff
@@ -0,0 +1,169 @@
+import time
+import six
+
+from google.cloud._helpers import _to_bytes
+from google.cloud.bigtable._generated import (
+    bigtable_pb2 as data_messages_v2_pb2)
+from google.gax import config, errors
+
+_MILLIS_PER_SECOND = 1000
+
+def _has_timeout_settings(backoff_settings):
+    return (backoff_settings.rpc_timeout_multiplier is not None and
+            backoff_settings.max_rpc_timeout_millis is not None and
+            backoff_settings.total_timeout_millis is not None and
+            backoff_settings.initial_rpc_timeout_millis is not None)
+
+class ReadRowsIterator():
+    """Creates an iterator equivalent to a_iter, but that retries on certain
+    exceptions.
+    """
+
+    def __init__(self, client, name, start_key, end_key, filter_, limit,
+        retry_options, **kwargs):
+        self.client = client
+        self.retry_options = retry_options
+        self.name = name
+        self.start_key = start_key
+        self.start_key_closed = True
+        self.end_key = end_key
+        self.filter_ = filter_
+        self.limit = limit
+
+        self.delay_mult = retry_options.backoff_settings.retry_delay_multiplier
+        self.max_delay_millis = retry_options.backoff_settings.max_retry_delay_millis
+        self.has_timeout_settings = _has_timeout_settings(retry_options.backoff_settings)
+
+        if self.has_timeout_settings:
+            self.timeout_mult = retry_options.backoff_settings.rpc_timeout_multiplier
+            self.max_timeout = (retry_options.backoff_settings.max_rpc_timeout_millis / _MILLIS_PER_SECOND)
+            self.total_timeout = (retry_options.backoff_settings.total_timeout_millis / _MILLIS_PER_SECOND)
+            self.set_stream()
+
+    def set_start_key(self, start_key):
+        self.start_key = start_key
+        self.start_key_closed = False
+
+    def set_stream(self):
+        request_pb = _create_row_request(
+            self.name, start_key=self.start_key,
+            start_key_closed=self.start_key_closed, end_key=self.end_key,
+            filter_= self.filter_, limit=self.limit)
+        self.stream = self.client._data_stub.ReadRows(request_pb)
+
+    def next(self, *args, **kwargs):
+        delay = self.retry_options.backoff_settings.initial_retry_delay_millis
+        exc = errors.RetryError('Retry total timeout exceeded before any'
+                                'response was received')
+        if self.has_timeout_settings:
+            timeout = (
+                self.retry_options.backoff_settings.initial_rpc_timeout_millis /
+                _MILLIS_PER_SECOND)
+
+            now = time.time()
+            deadline = now + self.total_timeout
+        else:
+            timeout = None
+            deadline = None
+
+        while deadline is None or now < deadline:
+            try:
+                return six.next(self.stream)
+            except StopIteration as stop:
+                raise stop
+            except Exception as exception:  # pylint: disable=broad-except
```

Comment:
Human-lint would like to know why we are catching every exception. :-)
Is it only possible to get GRPC exceptions here? If so, can we catch just those? If not, then this still will not work because line 77 expects a GRPC exception code.